### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by geoffrey-brier**
What I've done so far :
- added a new adapter (StreamHttpAdapter)
- fixed the way curl headers were set
- reduced the code dupplication
- added custom exception

### CODE REVIEW

**Comment by geoffrey-brier** [comment]
ping  @egeloen

**Comment by egeloen** [comment]
@geoffrey-brier What was the problem with the current curl implementation?

**Comment by geoffrey-brier** [comment]
They were badly handled

File: Model/StreamHttpAdapter.php [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    /**
+     * Calls an URL given a context.
+     *
+     * @param string   $url     A url
+     * @param resource $context A resource created from stream_context_create.
+     */
+    protected function call($url, $ctx)
+    {
+        if (($fp = fopen($this->fixUrl($url), 'rb', false, $ctx)) === false) {
+            throw new \Exception('An error occured when fetching the URL using streams.');
```
Comment (egeloen) [inline Model/StreamHttpAdapter.php]: You should rely on the `GoogleAnalyticsException`.

File: Model/StreamHttpAdapter.php [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        $content = stream_get_contents($fp);
+
+        fclose($fp);
+
+        if ($content === false) {
+            throw new \Exception('An error occured when fetching the URL using streams.');
```
Comment (egeloen) [inline Model/StreamHttpAdapter.php]: Same here & IMO the two messages should not be same.

File: Model/StreamHttpAdapter.php [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+        return $this->call($url, $ctx);
+    }
+
+    /**
+     * Calls an URL given a context.
+     *
+     * @param string   $url     A url
+     * @param resource $context A resource created from stream_context_create.
+     */
```
Comment (egeloen) [inline Model/StreamHttpAdapter.php:48]: Missing return.

File: Model/StreamHttpAdapter.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    public function getName()
+    {
+        return 'stream';
+    }
+
+    /**
+     * Fixes the URL (adds http:// if not set).
+     *
+     * @param string $url An URL.
+     */
```
Comment (egeloen) [inline Model/StreamHttpAdapter.php:76]: Missing return

File: Model/StreamHttpAdapter.php [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    /**
+     * Creates the stream context.
+     *
+     * @param string $method  The HTTP method (eg: GET, POST)
+     * @param array  $headers An array of headers.
+     * @param string $content The content.
+     *
+     * @return resource
```
Comment (egeloen) [inline Model/StreamHttpAdapter.php]: Missing description.

File: Exception/HttpAdapterException.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+     *
+     * @return \Widop\HttpAdapterBundle\Exception\HttpAdapterException An exception.
+     */
+    public static function invalidHeader($header, $info)
+    {
+        return new self(
+            sprintf('Invalid header %s (%s)', $header, $info)
+        );
+    }
+}
```
Comment (egeloen) [inline Exception/HttpAdapterException.php]: Missing empty line & you should put all extra information between double quotes as it is done in other bundle (ie. Cannot open URL "%s" with adapter "%s" (%s) & all exceptions should be suffixed by a dot :)

File: Model/BuzzHttpAdapter.php [hunk 7]
```diff
@@ -42,9 +44,18 @@ public function __construct(Browser $browser = null)
...
+        } catch (\Exception $e) {
+            throw HttpAdapterException::cannotFetchUrl(
+                $url,
+                $this->getName(),
+                $e->getMessage()
+            );
+        }
 
-        return $response->getContent();
+        return $content;
```
Comment (egeloen) [inline Model/BuzzHttpAdapter.php]: You should keep it as it was

File: Model/BuzzHttpAdapter.php [hunk 8]
```diff
@@ -52,9 +63,18 @@ public function getContent($url, array $headers = array())
...
+        } catch (\Exception $e) {
+            throw HttpAdapterException::cannotFetchUrl(
+                $url,
+                $this->getName(),
+                $e->getMessage()
+            );
+        }
 
-        return $reponse->getContent();
+        return $content;
```
Comment (egeloen) [inline Model/BuzzHttpAdapter.php]: Same here

File: Model/StreamHttpAdapter.php [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        // Rationalizes headers as an associative array
+        if (!empty($headers)) {
+            foreach ($headers as $key => $value) {
+                if (is_int($key)) {
+                    list ($fixedKey, $fixedValue) = $this->extractHeaderKeyAndValue($value);
+
+                    $rationalizedHeaders[$fixedKey] = $fixedValue;
+                } else {
+                    $rationalizedHeaders[$key] = trim($value);
+                }
```
Comment (egeloen) [inline Model/StreamHttpAdapter.php:105]: IMO, there is duplication code here. You should use something like:

``` php
if (is_int($key)) {
    list($key, $value) = $this->extractHeaderKeyAndValue($value);
}

$rationalizedHeaders[$key] = trim($value);
```

File: Model/StreamHttpAdapter.php [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            foreach ($rationalizedHeaders as $hKey => $hValue) {
+                $contextOptions['http']['header'] .= "$hKey:$hValue\r\n";
+            }
+        }
+
+        return stream_context_create($contextOptions);
+    }
+
+    /**
+     * Extract an header key ('Content-Length') and an header value ('42') from an
```
Comment (egeloen) [inline Model/StreamHttpAdapter.php]: Extract ->Extracts

File: Tests/Model/AbstractHttpAdapterTest.php [hunk 11]
```diff
@@ -29,6 +29,25 @@ protected function tearDown()
...
+     * @param string $methodName The method name.
+     *
+     * @return \ReflectionMethod A reflection method.
+     */
+    protected function getHttpAdapterReflectionMethod($methodName)
+    {
+        $method = new \ReflectionMethod(
+            get_class($this->httpAdapter),
+            $methodName
+        );
```
Comment (egeloen) [inline Tests/Model/AbstractHttpAdapterTest.php]: You should put it on one line.

File: Tests/Model/AbstractHttpAdapterTest.php [hunk 12]
```diff
@@ -38,7 +57,7 @@ public function testGetContentWithoutHeaders()
 
     public function testGetContentWithHeaders()
     {
-        $this->assertNotEmpty($this->httpAdapter->getContent('www.google.fr', array('Accept-Charset' => 'utf-8')));
+        $this->assertNotEmpty($this->httpAdapter->getContent('www.google.fr', array('Accept-Charset' => 'utf-8', 'Accept-Language: en-US,en;q=0.8')));
```
Comment (egeloen) [inline Tests/Model/AbstractHttpAdapterTest.php]: you should break this line.

File: Tests/Model/AbstractHttpAdapterTest.php [hunk 13]
```diff
@@ -51,27 +70,27 @@ public function testGetContentWithInvalidUrl()
...
     public function testPostContentWithoutHeaders()
     {
-        $this->assertNotEmpty($this->httpAdapter->postContent('www.google.fr'));
+        $this->assertNotEmpty($this->httpAdapter->postContent('www.widop.com'));
     }
 
     public function testPostContentWithHeaders()
     {
-        $this->assertNotEmpty($this->httpAdapter->postContent('www.google.fr', array('Accept-Charset' => 'utf-8')));
+        $this->assertNotEmpty($this->httpAdapter->postContent('www.widop.com', array('Accept-Charset' => 'utf-8', 'Accept-Language: en-US,en;q=0.8')));
```
Comment (egeloen) [inline Tests/Model/AbstractHttpAdapterTest.php]: you should break this line.

File: Tests/Model/CurlHttpAdapterTest.php [hunk 14]
```diff
@@ -32,4 +32,13 @@ public function testName()
...
         $this->assertSame('curl', $this->httpAdapter->getName());
     }
+
+    /**
+     * @expectedException \Widop\HttpAdapterBundle\Exception\HttpAdapterException
+     */
+    public function testExecuteThrowsInvalidCallbackException()
+    {
+        $this->getHttpAdapterReflectionMethod('execute')
+            ->invokeArgs($this->httpAdapter, array('foo', array(), '', 'tata'));
```
Comment (egeloen) [inline Tests/Model/CurlHttpAdapterTest.php:42]: tata :heart: :+1:

**Pull request merged**